cells/src/commonMain/kotlin/com/wire/kalium/cells/data/CellsDataSource.kt

```diff
@@ -0,0 +1,111 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ */
+package com.wire.kalium.cells.data
+
+import com.wire.kalium.cells.domain.CellsRepository
+import com.wire.kalium.cells.data.model.toDto
+import com.wire.kalium.cells.data.model.toModel
+import com.wire.kalium.cells.domain.model.CellNode
+import com.wire.kalium.cells.domain.model.PreCheckResult
+import com.wire.kalium.logic.NetworkFailure
+import com.wire.kalium.logic.functional.Either
+import com.wire.kalium.logic.functional.map
+import com.wire.kalium.network.utils.NetworkResponse
+import com.wire.kalium.util.KaliumDispatcher
+import com.wire.kalium.util.KaliumDispatcherImpl
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.withContext
+import okio.Path
+
+internal class CellsDataSource internal constructor(
+    private val cellsApi: CellsApi,
+    private val awsClient: CellsAwsClient,
+    private val dispatchers: KaliumDispatcher = KaliumDispatcherImpl
+) : CellsRepository {
+
+    override suspend fun preCheck(nodePath: String): Either<NetworkFailure, PreCheckResult> {
+        return withContext(dispatchers.io) {
+            wrapApiRequest {
+                cellsApi.preCheck(nodePath)
+            }.map { result ->
+                if (result.fileExists) {
+                    PreCheckResult.FileExists(result.nextPath ?: nodePath)
+                } else {
+                    PreCheckResult.Success
+                }
+            }
+        }
+    }
+
+    @Suppress("TooGenericExceptionCaught")
+    override suspend fun uploadFile(
+        path: Path,
+        node: CellNode,
+        onProgressUpdate: (Long) -> Unit,
+    ): Either<NetworkFailure, Unit> = withContext(dispatchers.io) {
+        try {
+            awsClient.upload(path, node.toDto(), onProgressUpdate)
+            Either.Right(Unit)
+        } catch (e: CancellationException) {
+            throw e
+        } catch (e: Exception) {
+            Either.Left(NetworkFailure.ServerMiscommunication(e))
+        }
+    }
+
+    override suspend fun getFiles(cellName: String): Either<NetworkFailure, List<CellNode>> =
+        withContext(dispatchers.io) {
+            wrapApiRequest {
+                cellsApi.getFiles(cellName)
+            }.map { response ->
+                response.nodes
+                    .filterNot { it.isRecycleBin }
+                    .map { it.toModel() }
+            }
+        }
+
+    override suspend fun deleteFile(node: CellNode): Either<NetworkFailure, Unit> {
+        return withContext(dispatchers.io) {
+            wrapApiRequest {
+                cellsApi.delete(node.toDto())
+            }
+        }
+    }
+
+    override suspend fun publishDraft(node: CellNode): Either<NetworkFailure, Unit> {
+        return withContext(dispatchers.io) {
```

Comment:
for api requests ktor will switch to io dispatchers internally so adding withContext here is redundent